### PR TITLE
docs: show red underlines in examples in rules docs

### LINKED ISSUE #16227: Change Request: Show red underlines in examples in rules docs
### ESLint version

N/A

### What problem do you want to solve?

* Make the example errors in rules docs easier to understand.
* Have the rules docs more directly demonstrate the user experience of enabling the rule (what you see in the docs mirrors what you will see in your editor)
* (optional) Surface actual lint error messages in the example code on hover.

### What do you think is the correct solution?

Implement a custom syntax highlighter (inspired by [shiki-twoslash]) which, at build time, optionally runs code through ESLint and augments the syntax-highlighted HTML with diagnostics. This could be implemented as code directly in ESLint, or as a standalone npm module pulled in as a dependency.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

I've been playing with a version of this locally and I have something which works, but it will need quite a bit more work to get it robust enough to be able to fully replace the existing Prism-based syntax highlighter.

Screen Shot 2022-08-22 at 2 04 57 PM

Thanks to @orta who has been looking into this with me.

### PR DESCRIPTION
**Pull request by ota-meshi**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

This PR modifies `docs/src/_plugins/md-syntax-highlighter.js` to indicate where an error occurs with a red marker in the examples in the rules document.

#### Is there anything you'd like reviewers to focus on?

close #16227

- This process is done using Prism's `after-tokenize` hook.
- ~~Content that starts with `/* eslint ` and whose language is js is considered an example of an eslint rule.~~
  It is now consider it an example of an ESLint rule only if it is wrapped in a custom container (that is, if parser options are passed from the custom container's rendering process).
  I would like to hear your opinion if this will work.
- I would like to hear your opinion on what style should be used.

### CODE REVIEW

**Comment by ota-meshi** [comment]
It seems that CI does not install root dependencies, which causes the documentation to fail to build.
Can you temporarily change to install root dependencies to see how this PR changes the documentation?

**Comment by ota-meshi** [comment]
It looks like this on my local.

image

**Comment by fasttime** [comment]
Thanks for this PR! CI is failing because the job that builds the docs now requires ESLint dependencies to be installed, but only deps in the docs package (in the subfolder) are installed at that point:

https://github.com/eslint/eslint/blob/1bf2520c4166aa55596417bf44c567555bc65fba/.github/workflows/docs-ci.yml#L34-L36

We could run `npm install` in the parent folder before running the `build` script, or we could move this particular job to the main workflow in .github/workflows/ci.yml and have it running from there. But in this case, maybe we'll need to run `npm install` in the docs package as well. You should be able do all changes by yourself, I think.

File: docs/package.json
Comment (ota-meshi) [inline docs/package.json:24]: I temporarily added postinstall so that I can check the results of this PR with netlify. I think you should probably replace this PR with another method before merging.
https://deploy-preview-18041--docs-eslint.netlify.app/rules/array-callback-return

File: docs/src/_plugins/prism-eslint-installer.js
Comment (ota-meshi) [reply docs/src/_plugins/prism-eslint-installer.js:60]: Thank you for your advice!
Currently, we are only using `Prism` hooks, so we cannot retrieve that information. We will probably need to work with `markdown-it`. I'll look into that method.

Comment (ota-meshi) [reply docs/src/_plugins/prism-eslint-installer.js]: I changed this and the documentation works fine now!
However, the test CI is no longer able to resolve `prismjs` dependencies, so the CI now fails 😓 I will look for a way to solve this problem.

Comment (ota-meshi) [reply docs/src/_plugins/prism-eslint-installer.js]: I refactored it and solved the problem in ci test 😎

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [inline docs/tools/prism-eslint-hook.js:74]: In some examples of the `no-multiple-empty-lines` rule we use the character `⏎` to indicate the presence of a line break, e.g.:

https://github.com/eslint/eslint/blob/5a8efd5b7ad13eb320a1f468d1d4ab3c8ab99214/docs/src/rules/no-multiple-empty-lines.md?plain=1#L61-L70

This character is only useful as a visual guide, it should not be parsed. In fact, it is removed from the code when the rule example is validated or when the Playground link is generated:

https://github.com/eslint/eslint/blob/5a8efd5b7ad13eb320a1f468d1d4ab3c8ab99214/docs/tools/markdown-it-rule-example.js#L75-L78

To get correct results, I think we should apply the same kind of transformation to the `code` that is being linted here.

The fact that the column number could be off by one after this transformation should not be an issue at all, because the `no-multiple-empty-lines` rule only ever reports on empty lines.

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js]: Thank you for checking this PR!
I fixed an issue related to `no-multiple-empty-lines`.
https://deploy-preview-18041--docs-eslint.netlify.app/rules/no-multiple-empty-lines

**Comment by nzakas** [comment]
I think it's okay to update the CI in this PR.

This looks really good, thanks! Can we change the red underline to look more like the one VS Code uses? (A little squiggly.) I'm pretty sure it's just a repeating image that we can find in the VS Code repo.

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [inline docs/tools/prism-eslint-hook.js:22]: I checked and we don't have any duplicate `content`, so this is okay. If we had any duplicates with different parser options, we could not rely just on the content as a key.

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [inline docs/tools/prism-eslint-hook.js:76]: Shall we merge this expression with [this one](https://github.com/eslint/eslint/blob/5a8efd5b7ad13eb320a1f468d1d4ab3c8ab99214/docs/tools/markdown-it-rule-example.js#L75-L78) into a separate function, to ensure that they don't get out of sync if we make a change?

Alternatively, we could extend `addContentMustBeMarked` to accept `code` as a new argument [here](https://github.com/eslint/eslint/blob/94a090719015448b7ede6204225734020f02b152/docs/.eleventy.js#L198-L200), so we could store it to use it later.

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [inline docs/tools/prism-eslint-hook.js:85]: I think it's fine to skip the fatals in `lines-around-comment`, but we should fix them in a separate PR.

File: docs/tools/prism-eslint-hook.js
Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js:22]: Prism.hooks only has code text in markdown, so I think it would be difficult to handle anything else as a key. For now, this is processed synchronously, so I don't think there is any problem with duplication. However, there may be problems after updating the document system 😓

File: docs/tools/prism-eslint-hook.js
Comment (amareshsm) [inline docs/tools/prism-eslint-hook.js:18]:
```suggestion
 * @param {string} content Source code content that marks ESlint errors.
```

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [reply docs/tools/prism-eslint-hook.js:23]: Every time execution hits this line, `contentsMustBeMarked` is empty, and every time execution hits the line with `contentsMustBeMarked.delete(env.code)` in the hook callback, the map has only one element: the content to be removed. So it seems that markdownIt is calling the highlighter immediately after calling the other plugin functions for the same block. If this is the case then we are safe: there can't be a duplication if there is just one potential match.

This makes me also think that probably we could just remember the last `content` and `options` found here instead of using a map. Do you think that would work?

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js:23]: > This makes me also think that probably we could just remember the last `content` and `options` found here instead of using a map. Do you think that would work?

That's right. I think it works. I would replace Map :+1:

**Comment by nzakas** [comment]
This is looking really nice. Is it possible to add a tooltip over the red-underlined code that includes the error message?

**Comment by ota-meshi** [comment]
> Is it possible to add a tooltip over the red-underlined code that includes the error message?

Hmm. It seems difficult to do that with Prism's hooks for now.
It doesn't seem possible to add anything other than the `<span>` class and content for each Token. (For example, we cannot add the `popovertarget` attribute only to the error token.)
If we request the Prism team to pass the Token instance to Prism's `wrap` hook, we might be able to get a tooltip, but I'm not sure yet 🤔

**Comment by ota-meshi** [comment]
> If we request the Prism team to pass the Token instance to Prism's wrap hook, we might be able to get a tooltip, but I'm not sure yet 🤔

I tried it but it didn't work well. During the `after-tokenize` hook and before, all Token instances created are replaced with new normalized Token instances before reaching the `wrap` hook.

**Comment by nzakas** [comment]
Is it possible to add a `title` attribute to the `<span>`?

**Comment by ota-meshi** [comment]
I don't know how to add `title` attribute (an attribute other than `class`) to a token using Prism's API 😓.
It might be possible to give a unique key to a `class` and later retrieve the message that maps to the unique key and set it to the `title`. But it may seem a bit like a hack.

**Comment by ota-meshi** [comment]
> It might be possible to give a unique key to a class and later retrieve the message that maps to the unique key and set it to the title.

I tried that. If someone doesn't like it that way, we can change it back.
I also made major changes to `prism-eslint-hook.js` because we needed to organize the token structure more to display the `title` properly.

image

https://deploy-preview-18041--docs-eslint.netlify.app/rules/array-callback-return

File: docs/package.json
Comment (nzakas) [reply docs/package.json:24]: We can add the top-level install to the workflow. :+1:

**Review by nzakas — COMMENTED** [review]
This looks amazing. :heart: 

We can update the docs CI to install the top-level packages and I think this is ready to go.

File: docs/package.json
Comment (ota-meshi) [reply docs/package.json]: I removed the `postinstall` and changed the workflow.
I think we probably need to modify the netlify script, but how should I do that?

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [reply docs/tools/prism-eslint-hook.js]: The fatals are fixed in #18078.

**Comment by fasttime** [comment]
Nice work! LGTM when the build is fixed.

File: docs/tools/prism-eslint-hook.js
Comment (mdjermanovic) [inline docs/tools/prism-eslint-hook.js:5]: How would this work for https://github.com/eslint/zh-hans.docs.eslint.org, where we don't have these modules?

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js:5]: Oh... I don't think it will work as it is now.
@kecrily Do you have any ideas how to solve it?

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js:5]: What do you think about changing this PR as below?

- `astUtils` is used only to get the regex, so it inlines the regex.
- Try `require("../../lib/api").Linter` and `require("eslint").Linter` and use the one that does not cause an error as the `Linter`.
- Change the install script of `zh-hans.docs.eslint.org` to install eslint as a dependency.

Comment (nzakas) [reply docs/tools/prism-eslint-hook.js:5]: Another option is to not do red underlines outside of the US site. We could check to see if these utilities exist, and if not, just don't do the linting.

Comment (snitin315) [reply docs/tools/prism-eslint-hook.js:5]: >Another option is to not do red underlines outside of the US site. We could check to see if these utilities exist, and if not, just don't do the lining.

I'd prefer this option.

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js]: > Another option is to not do red underlines outside of the US site. We could check to see if these utilities exist, and if not, just don't do the linting.

I've made changes to this PR to make it so 👍

**Comment by nzakas** [comment]
@christian-bromann looks like there's a problem with downloading the latest chromedriver.

File: docs/tools/prism-eslint-hook.js
Comment (nzakas) [inline docs/tools/prism-eslint-hook.js:10]: Let's add a comment here explaining why we're testing.
```suggestion
/*
 * We can only do the syntax highlighting in the English-language
 * site because we need the Linter and astUtils. This same
 * code is run in translation sites where these utilities are
 * not available, so check to see if they are before attempting
 * to highlight the code.
 */
try {
```

File: docs/tools/prism-eslint-hook.js
Comment (nzakas) [inline docs/tools/prism-eslint-hook.js:1]: Can you also add a fileoverview block on each new JS file?

**Comment by christian-bromann** [comment]
> looks like there's a problem with downloading the latest chromedriver.

There was, however we fixed it on our side so it should be resolved with the latest WebdriverIO version.

**Comment by nzakas** [comment]
I'll re-run now. Thanks!

**Review by nzakas — APPROVED** [review]
LGTM. Thanks! Would like another approval before merging.

**Comment by fasttime** [comment]
Since ESLint dependencies are no longer installed when creating the rule docs pages, red underlines don't show up any more in the deploy preview (e.g. https://deploy-preview-18041--docs-eslint.netlify.app/rules/no-const-assign). Is this what we want?

**Comment by ota-meshi** [comment]
I think we need to modify the netlify build script.
https://github.com/eslint/eslint/pull/18041#discussion_r1476814219

File: docs/package.json
Comment (nzakas) [reply docs/package.json]: Hmm. There is no Netlify script. It just checks out `docs` and does `npm install`. 

I wonder, could we just list `"eslint": "file:../"` in `docs/package.json` and that will work?

Comment (ota-meshi) [reply docs/package.json]: > I wonder, could we just list "eslint": "file:../" in docs/package.json and that will work?

I tried that locally but it doesn't seem to work. `eslint` dependencies were not installed 😓

Comment (ota-meshi) [reply docs/package.json]: Do you think putting the `netlify.toml` and adding the command config will work fine?
https://docs.netlify.com/configure-builds/file-based-configuration/#build-settings

Comment (nzakas) [reply docs/package.json]: Let's try it. I'm out of other ideas.

Comment (ota-meshi) [reply docs/package.json]: I tried adding `netlify.toml` and it seems to be working fine :tada:

https://deploy-preview-18041--docs-eslint.netlify.app/rules/array-callback-return

image

**Review by nzakas — APPROVED** [review]
Looks like everything is working now. :tada:

Would like @mdjermanovic to verify his changes before merging.

File: docs/tools/prism-eslint-hook.js
Comment (fasttime) [inline docs/tools/prism-eslint-hook.js:70]: In the Playground, the corresponding CSS class (`cm-lintPoint`) is also used to mark spans that are literally zero width, e.g. in this incorrect example for `unicode-bom` with option `always`: https://deploy-preview-18041--docs-eslint.netlify.app/rules/unicode-bom#always.

image

Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js:70]: Thank you for checking this PR!
I modified it so it looks the same as Playground :+1:

File: docs/src/assets/scss/syntax-highlighter.scss
Comment (fasttime) [inline docs/src/assets/scss/syntax-highlighter.scss:121]: Maybe we can move these styles into an `::after` pseudoelement, so they don't cause the rest of the line to shift rightwards?

Comment (ota-meshi) [reply docs/src/assets/scss/syntax-highlighter.scss:131]: I improved the style of the zero-width marker so that it appears more similar to playground 👍

Comment (fasttime) [reply docs/src/assets/scss/syntax-highlighter.scss]: It's looking good now, thanks!

**Review by fasttime — APPROVED** [review]

**Review by snitin315 — APPROVED** [review]
This looks great ⭐ , thanks!

Leaving it open for @mdjermanovic

File: docs/src/assets/scss/syntax-highlighter.scss
Comment (mdjermanovic) [inline docs/src/assets/scss/syntax-highlighter.scss:125]: For newlines, marks appear on the next line:

image

Maybe we should use `::before` instead of `::after`?

https://deploy-preview-18041--docs-eslint.netlify.app/rules/semi#always

Comment (ota-meshi) [reply docs/src/assets/scss/syntax-highlighter.scss]: Thank you for finding the problem! I fixed it.

File: docs/tools/prism-eslint-hook.js
Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js]: I changed this PR to make it look like VSCode by extracting the newline code character by character into tokens 👍

Comment (mdjermanovic) [reply docs/tools/prism-eslint-hook.js]: Would it be possible to do this only for newline characters that appear in an error range and are either at the very start of the range or lone on the line?

image

image

**Review by nzakas — APPROVED** [review]

**Comment by amareshsm** [comment]
nice 🎉

File: docs/tools/prism-eslint-hook.js
Comment (ota-meshi) [reply docs/tools/prism-eslint-hook.js]: I added CSS and changed it to look like that.

**Review by mdjermanovic — APPROVED** [review]
Great work, thanks!

**Pull request merged**
